tests/snappi/ecn/files/helper.py

```diff
@@ -0,0 +1,374 @@
+    data_flow = testbed_config.flows.flow(name=data_flow_name)[-1]
+    data_flow.tx_rx.port.tx_name = tx_port_name
+    data_flow.tx_rx.port.rx_name = rx_port_name
+
+    eth, ipv4 = data_flow.packet.ethernet().ipv4()
+    eth.src.value = tx_mac
+    eth.dst.value = rx_mac
+    eth.pfc_queue.value = prio
+
+    ipv4.src.value = tx_port_config.ip
+    ipv4.dst.value = rx_port_config.ip
+    ipv4.priority.choice = ipv4.priority.DSCP
+    ipv4.priority.dscp.phb.values = prio_dscp_map[prio]
+    ipv4.priority.dscp.ecn.value = (
+            ipv4.priority.dscp.ecn.CAPABLE_TRANSPORT_1)
+
+    data_flow.size.fixed = data_pkt_size
+    data_flow.rate.percentage = 100
+    data_flow.duration.fixed_packets.packets = data_pkt_cnt
+    data_flow.duration.fixed_packets.delay.nanoseconds = int(data_flow_delay_nanosec)
+
+    data_flow.metrics.enable = True
+    data_flow.metrics.loss = True
+    """ PFC Pause Storm """
+    pause_time = []
+    for x in range(8):
+        if x == prio:
+            pause_time.append(int('ffff', 16))
+        else:
+            pause_time.append(int('0000', 16))
+
+    vector = pfc_class_enable_vector([prio])
+
+    pause_flow = testbed_config.flows.flow(name=pause_flow_name)[-1]
+
+    """ Pause frames are sent from the RX port """
+    pause_flow.tx_rx.port.tx_name = testbed_config.ports[rx_port_id].name
+    pause_flow.tx_rx.port.rx_name = testbed_config.ports[tx_port_id].name
+
+    pause_pkt = pause_flow.packet.pfcpause()[-1]
+
+    pause_pkt.src.value = '00:00:fa:ce:fa:ce'
+    pause_pkt.dst.value = '01:80:C2:00:00:01'
+    pause_pkt.class_enable_vector.value = vector
+    pause_pkt.pause_class_0.value = pause_time[0]
+    pause_pkt.pause_class_1.value = pause_time[1]
+    pause_pkt.pause_class_2.value = pause_time[2]
+    pause_pkt.pause_class_3.value = pause_time[3]
+    pause_pkt.pause_class_4.value = pause_time[4]
+    pause_pkt.pause_class_5.value = pause_time[5]
+    pause_pkt.pause_class_6.value = pause_time[6]
+    pause_pkt.pause_class_7.value = pause_time[7]
+
+    speed_str = testbed_config.layer1[0].speed
+    speed_gbps = int(speed_str.split('_')[1])
+    pause_dur = 65535 * 64 * 8.0 / (speed_gbps * 1e9)
+    pps = int(2 / pause_dur)
+
+    pause_flow.rate.pps = pps
+    pause_flow.size.fixed = 64
+    pause_flow.duration.fixed_seconds.seconds = exp_dur_sec
+    pause_flow.duration.fixed_seconds.delay.nanoseconds = 0
+
+    pause_flow.metrics.enable = True
+    pause_flow.metrics.loss = True
+
+
+def __config_capture_ip_pkt(testbed_config, port_id):
+    """
+    Generate the configuration to capture IP packets
+
+    Args:
+        testbed_config (obj): L2/L3 config of a T0 testbed
+        port_id (int): ID of DUT port to capture packets
+
+    Returns:
+        N/A
+    """
+
+    """ We only capture IP packets """
+    cap = testbed_config.captures.capture(name='rx_capture')[-1]
+    cap.port_names = [testbed_config.ports[port_id].name]
+    cap.format = cap.PCAP
+
+    ip_filter, = cap.filters.filter()
```

Comment:
What does this filter mean? Can you add some comments?

Reply:
Why is 'ip_filter,' rather than 'ip_filter'

Reply:
Left the variable(ip_filter) as same in https://github.com/Azure/sonic-mgmt/blob/master/tests/ixia/ecn/files/helper.py#L283. Please let me know if you want me to change the variable name

Reply:
I wonder know why it is **ip_filter, = cap.filters.filter()** rather than **ip_filter = cap.filters.filter()**. In addition, I wonder know the meaning of '40', '14' , and '0F'. Do you mean byte[14] & 0x0f should be '40'?

Reply:
Sorry... I misread your comment, filters is an array...I have updated <b>ip_filter, = cap.filters.filter()</b> to  <b>ip_filter = cap.filters.filter()[-1]</b> for linearity.

![image](https://user-images.githubusercontent.com/48308607/130568966-56bf6926-cfbf-4290-90d8-912c29a2f6c3.png)

<b>`
Do you mean byte[14] & 0x0f should be '40'? Yes, you are correct. Value is hex '40' & wildcard mask is  0x0f`</b>

![image](https://user-images.githubusercontent.com/48308607/130569894-aeb4aaf5-ce83-4aaa-a488-c92db832f663.png)


Models:
=====
https://redocly.github.io/redoc/?url=https://raw.githubusercontent.com/open-traffic-generator/models/master/artifacts/openapi.yaml

Reply:
byte[14] should be first byte of IP header (first four bits should be 0x4, and last four bits should be IP header length in 32-bit words). Hence, byte[14] & 0x0f should be IP header length in 32-bit words.

Reply:
byte[14] should be first byte of IP header (first four bits should be 0x4, and last four bits should be IP header length in 32-bit words). Since 0x0f is a wildcard mask the capture filters ipv4 packets based on version.

![image](https://user-images.githubusercontent.com/48308607/130578209-f28b5a65-869b-488c-9730-cebbe2606382.png)

Reply:
So it is actually byte[14] & (~mask) = byte[14] & 0xF0 = 0x40, right?

Reply:
Yes..right